docs/specs/app-initiated-sso-to-web.md

```diff
@@ -0,0 +1,179 @@
+- [Token Exchange Profile for App Initiated SSO to Web](#token-exchange-profile-for-app-initiated-sso-to-web)
+  - [Defined token type identifiers](#defined-token-type-identifiers)
+  - [Token Exchange Request](#token-exchange-request)
+  - [Token Exchange Response](#token-exchange-response)
+- [Exchanging for browser cookie of a session with the authorization endpoint](#exchanging-for-browser-cookie-of-a-session-with-the-authorization-endpoint)
+- [Security Considerations](#security-considerations)
+  - [Binding Tokens To Device](#binding-tokens-to-device)
+- [Changes In SDKs](#changes-in-sdks)
+- [Usecases](#usecases)
+
+## Abstract
+
+This document describes a mechanism that allows a cookie based web app to share the authentication obtained by a mobile app.
+
+This document is drafted based on [OpenID Connect Native SSO for Mobile Apps 1.0](https://openid.net/specs/openid-connect-native-sso-1_0.html).
+
+## The Flow
+
+```mermaid
+sequenceDiagram
+    participant app1 as Mobile App #35;1
+    participant Web as Web App #35;2 <br/>(Cookie Based)<br/>https://www.example.com
+    participant Authgear as Authgear <br/> https://auth.example.com
+    app1->>+Authgear: [1]GET https://auth.example.com/oauth2/authorize<br/>client_id=1<br/>scope=x_device_browser_sso<br/>redirect_uri=app://redirect
+    Note right of Authgear: Alice enter user credentials
+    Authgear->>-app1: [2]Redirect to app://redirect?code=CODE
+    app1->>+Authgear: [3]POST https://auth.example.com/oauth2/token<br/>code=CODE
+    Authgear->>-app1: [4]Response with<br/>access_token=ACCESS_TOKEN<br/>refresh_token=REFRESH_TOKEN<br/>id_token=ID_TOKEN<br/>device_secret=DEVICE_SECRET
+    app1->>+Authgear: [5]POST https://auth.example.com/oauth2/token<br/>grant_type=token-exchange<br/>client_id=2<br/>subject_token_type=urn:ietf:params:oauth:token-type:id_token<br/>subject_token=ID_TOKEN<br/>actor_token_type=urn:x-oath:params:oauth:token-type:device-secret<br/>actor_token=DEVICE_SECRET<br/>requested_token_type=urn:x-oath:params:oauth:token-type:device-browser-session-token
+    Authgear->>-app1: [6]Response with<br/>device_browser_session_token=DEVICE_BROWER_SESSION_TOKEN
+    app1->>Authgear: [7]Open URL In Browser: https://auth.example.com/oauth2/authorize?<br/>client_id=2<br/>id_token_hint=ID_TOKEN<br/>device_browser_session_token=DEVICE_BROWER_SESSION_TOKEN<br/>redirect_uri=https://www.example.com/redirect<br/>state=STATE<br/>prompt=none<br/>response_type=token<br/>response_mode=cookie
+    Authgear->>Web: [8]Redirect to https://www.example.com/redirect?state=STATE<br/>Set Cookie at=ACCESS_TOKEN
+    Note right of Web: Web now authenticated as Alice by cookie
+```
+
+Step [1] to [4] are the standard OpenID Connect authorization_code flow with the following extensions. In step [1], the `x_device_browser_sso` is specified signifying that the client is requesting a `device_secret` to be returned when the code is exchanged for tokens.
```

Comment:
Remember to rename all ocurrences of `x_device_browser_sso`

Reply:
Are we going to rename it to `x_app_initiated_sso_to_web`?

Reply:
Let me name it `x_device_browser_session_token` for now

Reply:
When we correct our Token Response, we include the token in the key `access_token`.

The only remaining place `x_device_browser_session_token` will appear is in the authorization endpoint. When I think it again, it is not a good practice to have side effects in GET request. It is more appropriate to have side effects in POST request.

Ideally, we should perform two token exchanges.

1. Token Exchange 1: id token + device secret -> a short live token
2. Token Exchange 2: the short live token -> cookie

Token Exchange 2 has to happen in the webview. But this design is not compatible with ASWebAuthenticationSession. But if we assume App-initiated-sso-to-web is only available for webview, then it is not a problem. We can offer a method in the SDK to return a script to be run in the webview to perform Token Exchange 2.

Reply:
>  When I think it again, it is not a good practice to have side effects in GET request.

However, the /authorize endpoint always has side effects such as setting oauth session cookies. I think setting the device browser session in this endpoint is more correct.
And actually `/token` never handles cookies.

> But if we assume App-initiated-sso-to-web is only available for webview

I think this assumption may not be correct, as usecase 1 probably want to open the marketing site in a normal browser.